design/src/docs/anatomy.md

```diff
@@ -0,0 +1,687 @@
+    }
+}
+```
+
+We go into more detail on how the `Upgradable` trait is used in conjunction with builders in the [Builders](#builders) section below.
+
+## Routers
+
+Different protocols supported by Smithy enjoy different routing mechanisms, for example, [AWS JSON 1.0](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-json-1_0-protocol.html#protocol-behaviors) uses the `X-Amz-Target` header to select an operation, whereas [AWS REST XML](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-restxml-protocol.html) uses the [HTTP label trait](https://awslabs.github.io/smithy/2.0/spec/http-bindings.html#httplabel-trait).
+
+Despite their differences, all routing mechanisms satisfy a common interface. This is formalized using the `Router` trait:
+
+```rust
+/// An interface for retrieving an inner [`Service`] given a [`http::Request`].
+pub trait Router {
+    type Service;
+    type Error;
+
+    /// Matches a [`http::Request`] to a target [`Service`].
+    fn match_route(&self, request: &http::Request) -> Result<Self::Service, Self::Error>;
+}
+```
+
+which provides the ability to lookup an inner HTTP service from a collection using a `&http::Request`.
+
+Types which implement the `Router` trait are converted to a HTTP service via the `RoutingService` struct:
+
+```rust
+/// A [`Service`] using a [`Router`] `R` to redirect messages to specific routes.
+///
+/// The `Protocol` parameter is used to determine the serialization of errors.
+pub struct RoutingService<R, Protocol> {
+    router: R,
+    _protocol: PhantomData<Protocol>,
+}
+
+impl<R, P> Service<http::Request> for RoutingService<R, P>
+where
+    R: Router<B>,
+    R::Service: Service<http::Request, Response = http::Response>,
+    R::Error: IntoResponse<P> + Error,
+{
+    type Response = http::Response;
+    type Error = Box<dyn Error>;
+
+    async fn call(&mut self, req: http::Request<B>) -> Result<Self::Response, Self::Error> {
+        match self.router.match_route(&req) {
+            // Successfully routed, use the routes `Service::call`.
+            Ok(ok) => ok.oneshot(req).await,
+            // Failed to route, use the `R::Error`s `IntoResponse<P>`.
+            Err(error) => {
+                debug!(%error, "failed to route");
+                Err(Box::new(error.into_response()))
+            }
+        }
+    }
+}
+```
+
+The `RouterService` is the final piece necessary to form a functioning composition - it is used to aggregate together the HTTP services, created via the upgrade procedure, into a single HTTP service which can be presented to the customer.
+
+![RouterService](imgs/router.png)
+
+## Builders
+
+The service builder is the primary public API, it can be generated for every [Smithy Service](https://awslabs.github.io/smithy/2.0/spec/service-types.html). At a high-level, the purpose of a service builder is to take a function for each Smithy Operation, whose signature is constrained by the Smithy model, and produce a single HTTP service.
```

Comment:
Or closure, or model service, as explained in the beginning.

Reply:
I'm tempted to leave it as is because a [`Service`](https://docs.rs/tower-service/latest/tower_service/trait.Service.html) is already described as a

>An asynchronous function from a Request to a Response.

And closures are anonymous functions, so this is "true enough".